types/hooks.d.ts

```diff
@@ -0,0 +1,235 @@
+    values: UseSpringProps<DS>
+  ): ForwardedProps<DS>
+
+  // there's a third value in the tuple but it's not public API (?)
+  export function useTrail<DS extends object>(
+    count: number,
+    getProps: () => UseSpringProps<DS>
+  ): [ForwardedProps<DS>[], SetUpdateFn<DS>]
+  export function useTrail<DS extends object>(
+    count: number,
+    values: UseSpringProps<DS>
+  ): ForwardedProps<DS>[] // safe to modify (result of .map)
+
+  interface UseTransitionProps<TItem, DS extends object>
+    extends HooksBaseProps {
+    items: ReadonlyArray<TItem> | null | undefined
+    /**
+     * Spring config, or for individual items: fn(item => config)
+     * @default config.default
+     */
+    config?: SpringConfig | ((item: TItem) => SpringConfig)
+    /**
+     * The same keys you would normally hand over to React in a list. Keys can be specified as a key-accessor function, an array of keys, or a single value
+     */
+    keys?:
+      | ((item: TItem) => TransitionKeyProps)
+      | ReadonlyArray<TransitionKeyProps>
+      | TransitionKeyProps
+
+    /**
+     * When true enforces that an item can only occur once instead of allowing two or more items with the same key to co-exist in a stack
+     * @default false
+     */
+    unique?: boolean
+    /**
+     * Trailing delay in ms
+     */
+    trail?: number
+
+    from?: InferFrom<DS>
+    /**
+     * Values that apply to new elements, or: item => values
+     * @default {}
+     */
+    enter?: InferFrom<DS> | ((item: TItem) => InferFrom<DS>)
+    /**
+     * Values that apply to leaving elements, or: item => values
+     * @default {}
+     */
+    leave?: InferFrom<DS> | ((item: TItem) => InferFrom<DS>)
+    /**
+     * Values that apply to elements that are neither entering nor leaving (you can use this to update present elements), or: item => values
+     */
+    update?: InferFrom<DS> | ((item: TItem) => InferFrom<DS>)
+  }
+
+  type TransitionState = 'enter' | 'leave' | 'update'
+
+  interface UseTransitionResult<TItem, DS extends object> {
+    item: TItem
+    key: string
+    state: TransitionState
+    props: ForwardedProps<DS>
+  }
+
+  export function useTransition<TItem, DS extends object>(
+    values: Merge<DS, UseTransitionProps<TItem, DS>>
+  ): UseTransitionResult<TItem, ForwardedProps<DS>>[] // result array is safe to modify
+
+  // higher order hooks 🤯
+  export namespace useKeyframes {
+    // this kind of higher order hook requires higher kinded types to type correctly at all
+    // this is the best approximation I can get to....... and there sure are anys around 🤯
+
+    // the docs says it receives a 3rd argument with the component's own props, but it's only
+    // ever called with two (https://github.com/react-spring/react-spring/blob/v7.1.3/src/hooks/KeyframesHook.js#L25)
+    export type KeyframeFn<DS extends object> = (
+      next: (props: UseSpringProps<DS>) => Promise<void>,
+      cancel: () => void
+    ) => Promise<void>
```

Comment:
I tried following around to see if `Controller` or `useSpring` did something that would effectively give that 3rd argument but I didn't see anything.

Documentation mistake / copy&paste from the `Keyframe` class?